### PR TITLE
[FLINK-34494][table] Migrate ReplaceIntersectWithSemiJoinRule to java.

### PR DESCRIPTION
**Pull request by liuyongvs**
## What is the purpose of the change

The PR migrates ReplaceIntersectWithSemiJoinRule to java
it doesn't touch ReplaceIntersectWithSemiJoinRuleTest to be sure that java version continues passing it

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by liuyongvs** [comment]
hi @dawidwys will you also help review this pr ?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.calcite.util.Util;
+import org.immutables.value.Value;
+
+import java.util.List;
+
+import static org.apache.flink.table.planner.plan.utils.SetOpRewriteUtil.generateEqualsCondition;
+
+/**
+ * Planner rule that replaces distinct {@link org.apache.calcite.rel.core.Intersect} with a distinct
+ * {@link org.apache.calcite.rel.core.Aggregate} on a SEMI {@link org.apache.calcite.rel.core.Join}.
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java:38]: nit: is there a reason we use absolute name in javadocs?
in scala it was relative why can't we follow same approach?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    /** Rule configuration. */
+    @Value.Immutable(singleton = false)
+    public interface ReplaceIntersectWithSemiJoinRuleConfig extends RelRule.Config {
+        ReplaceIntersectWithSemiJoinRule.ReplaceIntersectWithSemiJoinRuleConfig DEFAULT =
+                ImmutableReplaceIntersectWithSemiJoinRule.ReplaceIntersectWithSemiJoinRuleConfig
+                        .builder()
+                        .build()
+                        .withOperandSupplier(b0 -> b0.operand(Intersect.class).anyInputs())
+                        .withRelBuilderFactory(RelFactories.LOGICAL_BUILDER)
+                        .withDescription("ReplaceIntersectWithSemiJoinRule");
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java:89]: very nit: in fact could be left as is
however it also could be moved to builder and then the will not be created extra objects

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java (hunk 1 above)
Comment (liuyongvs) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java]: @snuyanzin fixed

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java (hunk 2 above)
Comment (liuyongvs) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/ReplaceIntersectWithSemiJoinRule.java:91]: left as is, others do it same

**Review by snuyanzin — APPROVED** [review]
LGTM

**Pull request merged**